### DISCUSSION TITLE
Investigating reliability

### DISCUSSION THREAD

**Discussion by ccrvlh** [rq/rq#1756] [General]
There have been some reports of missed jobs issue #758 have quite a few experiences with this. Even though this has been reported by a lot of different people, we haven't been able to replicate it yet. I gathered what I considered the most relevant cues on what issue might be.

https://github.com/rq/rq/issues/758#issue-189610691
Job doesn’t seem to be picked up from the queue - implies an issue with the worker
Heartbeat to prevent worker timeout are also not shown.

https://github.com/rq/rq/issues/758#issuecomment-306677675
Also implies issue with the worker: job stays as “queued” but worker doesn’t dequeue it. In this case, it gets tricky since another job sent to same queue, same worker worked.

https://github.com/rq/rq/issues/758#issuecomment-340325037
Possible fix by reducing `default_worker_ttl` to 120 although another comment says this didn’t fix it.

https://github.com/rq/rq/issues/758#issuecomment-536110780
This issue mentions the Queue does not have any reference to the job, so even though the job is sent to redis, is not being persisted to the queue

https://github.com/rq/rq/issues/758#issuecomment-573882950
This implies that running the worker with `multiprocessing.Process` fixes the issue

https://github.com/rq/rq/issues/758#issuecomment-614933609
Potential fix by requeueing the job.

https://github.com/rq/rq/issues/758#issuecomment-773583998
Implies the issue goes to `rq:job` but doesn’t go to the `rq:queues:xyz`.
Apparently `queue.jobs` doesn’t show jobs but `job_class.fetch(id)` does. Which would be a similar issue to the one reported here https://github.com/rq/rq/issues/758#issuecomment-536110780

https://github.com/rq/rq/issues/758#issuecomment-773649573
Potential partial failure on the `pipeline` could explain? Check whether this is already within a transaction, evaluate using Lua to improve reliability here.

https://github.com/rq/rq/issues/758#issuecomment-773663509
Job might get lost after is popped from the queue before is added to StartedJobRegistry

https://github.com/rq/rq/issues/758#issuecomment-773699687
> It looks like jobs that get stuck in this queued state call `rpush` to add the job to the Redis queue, but they don’t get popped off the queue in the worker.
Implies an issue in the `lpop` method.

https://github.com/rq/rq/issues/758#issuecomment-774252808
Potential effect of multiple workers, reinforce hypothesis that the issue is on the `lpop` method.

https://github.com/rq/rq/issues/758#issuecomment-774531964
Potential workaround by having workers suspended when the job is sent to the queue

https://github.com/rq/rq/issues/758#issuecomment-776196743
Potential Redis update to 5.0.6 fixed it?

https://github.com/rq/rq/issues/758#issuecomment-791603930
Mentions potential Redis load being related to the issue.

[Job queued but not run · Issue #758 · rq/rq · GitHub](https://github.com/rq/rq/issues/758#issuecomment-793495443)
Implies that speed at which jobs are sent may increase likelihood of lost jobs.

[Job queued but not run · Issue #758 · rq/rq · GitHub](https://github.com/rq/rq/issues/758#issuecomment-803046098)
Implies that the `SimpleWorker` is more susceptible to this issue.

https://github.com/rq/rq/issues/758#issuecomment-833207507
Implies potential issue with  `blpop`  (maybe only with Redis-Py?) silently hanging forever, can be addressed by the `socket_timeout`. This is consistent with a few other comments.

https://github.com/rq/rq/issues/758#issuecomment-1383190027
Implies that issue occurs after the `execute_job` method is called, but  before the `prepare_for_execution`

 So far, best hypothesis imho is an issue with the `blpop` command hanging forever. If this is indeed the case (couldn't be replicated), a potential fix would include setting a `socket_timeout` on the connection, to force a timeout at a lower level and then handle the timeout with a retry logic. This hypothesis is not compatible to the symptom where we can't get see the job on the `rq:queue:<job_id>` though, since that would mean that the there is a bug in the `get_job_ids` method and maybe the issue was at the `push_job_ids` method, meaning the job was not being pushed to the queue.

**Comment by OlegZv** [comment]
We're experiencing some issues with reliability. Just opened a PR to a potentially unrelated issue #1764. But still want to describe what we're seeing.

Setup: running the workers inside a Kubernetes pod with `supervisord`. The Dockerfile base image is `python:3.10.4-slim-buster`. Redis `redis_version:6.0.14`. RQ version `1.8.1`, but the issue persists in `1.12.0`. 99% of the application is utilizing the scheduling functionality.

Symptoms: some workers just end up hanging. The `supervisorctl` says the process is `RUNNING`. But none of the queue items are being processed.

Logs: the issue definitely happens around the restart of the Redis cluster (done by the cloud provider)
Reproducibility: could only reproduce a dying scheduler process locally, that's what the #1764 is for. So far  the worker managed to survive server being unavailable and me killing the connections of the worker.
Redis note: while killing the connection to see if that will make the worker to hang I noticed that the `subscribe` feature of the worker would be dead and not re-subscribe after killing the subscribe connection.
Debugging: unfortunately the k8s doesn't allow `ptrace` by default. Enabling it this week and hope the hanging worker happens again.

**Reply by ccrvlh** [reply]
@OlegZv thanks for sharing this. This does look somewhat related to [this discussion](https://github.com/rq/rq/issues/607#issuecomment-159338168), where there's a confusion with `supervisord` not being able to recognize the death of a worker. Did you try adding a `socket_timeout` to your Redis connection? I understand this is not entirely related to this specific issue, but it should help with a loud exception - wondering if it would change anything, meaning: if the scheduler had a `TimeoutException` would supervisor recognize the process as dead and start the new one?

Also when none of the queues items were being processed, did you wait for the fallback period? This is currently set at 10 minutes, and it has recently been changed to two minutes in PR #1759 

Last edit, also related to that first issue. I'm not entirely sure since I've never looked in that, but the scheduler runs in a `Process`, maybe `supervisord` is tracking the parent process, so it doesn't realize the child process running the `scheduler` died. I don't usually use supervisord, I normally just run workers on pods directly, and I do have a mechanism to refresh workers periodically, so never experienced that.

**Reply by OlegZv** [reply]
Re: `supervisord` and `scheduler` - the supervisord definitely doesn't track the `scheduler` that's why most of the time our scheduler is dead, but the queue worker might still be working

Re: fallback period -  yeah, the workers are hanging and schedulers are down on the `dev` server for at least two days now. Keeping pods active in case I need to debug anything.

Re: TimeoutException - I might try this if I can't debug a live "stuck" worker. But I've already seen some workers throwing the `TimeoutException` from time to time when the server restarts, but it might be that they were in a different redis command (`subscribe` handler vs. `lpop`)

**Reply by ccrvlh** [reply]
Yeah, makes sense. So we can exclude the fallback period.

It seems that the issue is with the `Process` implementation indeed. In that other thread folks were mentioning the child process having the same name (which was supposed to be a PR that never came). If I understand it all correctly, the weird thing here is that killing the scheduler doesn't actually kill both processes (the parent / origin process, and the child process in which the `scheduler` is actually running), which causes the `supervisord` to not function correctly.

I would still try the `socket_timeout` though, if everything goes right it shouldn't make any difference (the `socket_timout` should be higher than the `ttl`), but if things goes south the `timeout` at the socket level would help to understand. The last couple of commits on the master branch also should help with debugging - it won't make much of a difference in your case, since the issue is with the scheduler, but just so you know.

**Reply by OlegZv** [reply]
@lowercase00 I'm wondering, what would happen if the Redis server restarted and wiped all the data on the server? I know that the `subscribe` functionality would no longer work on the worker side. The scheduler process is _likely_ to die due to the TimeoutException. Not sure how the worker process would react to that though.

**Reply by ccrvlh** [reply]
I'm not really sure to be honest, would have to run some tests. I think as is, the socket would timeout, but I think pods would get recycled and then fall into the connection cycle. There are a couple of opened PRs that improves connection handling, but it was decided not to include a retry with backoff logic on the timeout just yet, but definitely something to improve for the future. As this would happen in the main worker process (before the fork I mean), I think `supervisord` or any other process manager would handle it just fine (with a restart). But I'm not really sure.
